src/waveform/renderers/waveformmarkset.h

```diff
@@ -19,17 +20,63 @@ class WaveformMarkSet {
                const SkinContext& context,
                const WaveformSignalColors& signalColors);
 
-    inline QList<WaveformMarkPointer>::const_iterator begin() const { return m_marks.begin(); }
-    inline QList<WaveformMarkPointer>::const_iterator end() const { return m_marks.end(); }
+    template<typename Receiver, typename Slot>
+    void connectSamplePositionChanged(Receiver receiver, Slot slot) const {
+        for (const auto& pMark : std::as_const(m_marks)) {
+            if (pMark->isValid()) {
+                pMark->connectSamplePositionChanged(receiver, slot);
+            }
+        }
+    };
+
+    template<typename Receiver, typename Slot>
+    void connectSampleEndPositionChanged(Receiver receiver, Slot slot) const {
+        for (const auto& pMark : std::as_const(m_marks)) {
+            if (pMark->isValid()) {
+                pMark->connectSampleEndPositionChanged(receiver, slot);
+            }
+        }
+    };
+
+    template<typename Receiver, typename Slot>
+    void connectVisibleChanged(Receiver receiver, Slot slot) const {
+        for (const auto& pMark : std::as_const(m_marks)) {
+            if (pMark->hasVisible()) {
+                pMark->connectVisibleChanged(receiver, slot);
+            }
+        }
+    }
+
+    inline QList<WaveformMarkPointer>::const_iterator begin() const {
+        return m_marksToRender.begin();
+    }
+    inline QList<WaveformMarkPointer>::const_iterator end() const {
+        return m_marksToRender.end();
+    }
+    inline QList<WaveformMarkPointer>::const_iterator cbegin() const {
+        return m_marksToRender.cbegin();
+    }
+    inline QList<WaveformMarkPointer>::const_iterator cend() const {
+        return m_marksToRender.cend();
+    }
 
     // hotCue must be valid (>= 0 and < NUM_HOT_CUES)
     WaveformMarkPointer getHotCueMark(int hotCue) const;
     WaveformMarkPointer getDefaultMark() const;
+    WaveformMarkPointer findHoveredMark(QPoint point, Qt::Orientation orientation) const;
+
+    void update();
 
   private:
-    void clear() { m_marks.clear(); }
+    void clear() {
+        m_marks.clear();
+        m_marksToRender.clear();
+    }
     WaveformMarkPointer m_pDefaultMark;
     QList<WaveformMarkPointer> m_marks;
+    // List of visible WaveformMarks sorted by the order they appear in the track
+    QList<WaveformMarkPointer> m_marksToRender;
+
     QMap<int, WaveformMarkPointer> m_hotCueMarks;
```

Comment:
we now have three different containers referencing the same marks. That seems excessive. Can you do me the favor and take another look at waveformmarkset and see if there is potential for simplifying the class?

Reply:
It sure does feel excessive but I didn't add any, I only moved m_marksToRender from outside (WOverview) to inside.

Reply:
There absolutely is potential to simplifying and improving WaveformMarkSet and actually I went down the rabbit hole of the TODO I added. But this PR is the minimum refactoring I felt that could go into 2.4 as it actually fixes two bugs. Anything beyond that comes with too much risk IMO.